### PR TITLE
Programmatic UniProt search

### PR DESCRIPTION
**Pull request by Will-Tyler**
### Acknowledgements

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### Description

This pull request adds functionality to search the UniProt database programmatically using the [UniProt query API](https://www.uniprot.org/help/api_queries). I update the tutorial to demonstrate how to use the new functionality, and I update the NEWS file to mention the new functionality.

Here is a basic example of the functionality:

```python

from Bio.UniProt import UniProt

query = "(organism_id:2697049) AND (reviewed:true)"
results = list(UniProt.search(query))
```

### Testing

I added unit tests that do some basic checks. These unit tests require an internet connection. I ran the tests locally and they passed.

For the documentation changes (the tutorial and the API docs), I manually built and inspected the documentation.

I manually tested the examples in the tutorial in the Python console.

### CODE REVIEW

**Comment by Will-Tyler** [comment]
UniProt's [REST API documentation](https://www.uniprot.org/help/api) does mention that users can specify an email address to contact in case there are issues while using the API. (I am not exactly sure what the format to provide the email is, but I can try to figure this out.) We can consider requiring an email from the user like how the Entrez module does.

**Comment by Will-Tyler** [comment]
I think the macOS CI test failed due to a transient error while connecting to GitHub...

**Comment by peterjc** [comment]
To discuss: Rather than ``Bio/UniProt/UniProt.py`` which is repetitive, and mixed case (see PEP8 but note we have historical issues with module names), maybe something more like ``Bio/UniProt/web_api.py``? Or stick the code in ``Bio/UniProt/__init__.py``?

**Comment by peterjc** [comment]
If this was a large bit of code, then a separate file would seem best. But for a couple of public functions only I'm OK with putting them in ``__init__.py``.

File: Bio/UniProt/__init__.py [hunk 1]
```diff
@@ -15,3 +15,106 @@
...
+        except StopIteration:
+            if self.next_url is not None:
+                with urlopen(self.next_url) as response:
+                    self.next_url = _get_next_link(response.headers)
+                    self.batch_iterator = iter(
+                        json.loads(response.read().decode())["results"]
+                    )
+                    return next(self.batch_iterator)
+            else:
+                raise StopIteration
```
Comment (mdehoon) [inline Bio/UniProt/__init__.py:65]: Does the response headers tell you how many results are returned? If so, we can add a `__len__` method to this iterator, returning the number of results without having to actually iterate over them.
Also, we may want to add a `total` or `total_count` attribute to show that total number of queries that would have been obtained if we had not limited `size` to 500.
If we want to be fancy, we could use the `cursor` argument to obtain the results on demand. Then we won't need a `count_search_results_for` function. In that case, `__len__` should return the total number of queries obtainable. Perhaps we want to do that later, and then I am a bit uneasy to include a `count_search_results_for` function now if we don't need it in the end (and will deprecate it).

**Comment by Will-Tyler** [comment]
Is this good to merge once I resolve the merge conflict?

**Comment by mdehoon** [comment]
@Will-Tyler  Thank you, this is looking good.

Just one thing I was thinking about:

Usually an iterator is used if there are too many items to be stored in a list or tuple.
Here, the situation is a bit different: There is no problem storing all the items, but downloading them is expensive.
Then, instead of an iterator, something more like a lazy-loading list may make more sense.
I think just adding a ``__getitem__`` method to `_UniProtSearchResultIterator` may be sufficient. This method will look in a cache to see if the requested record(s) was already downloaded, and if not, download it from UniProt.
For a user, the result of `UniProt.search` will then effectively act like a list, like this:

```python
>>> from Bio import UniProt
>>> query = "(organism_id:2697049) AND (reviewed:true)"
>>> results = UniProt.search(query)
>>> len(results)
1234
>>> top_records = results[:10]  # this downloads the first ten records
>>> top_record = results[0]  # this returns the top record from the cache
```

`results` can still act as an iterator, taking records from the same cache, or downloading them if needed.

What do you think?

**Comment by Will-Tyler** [comment]
I looked into adding a `__getitem__` method today. I think that slicing is subtlely complicated. We want to mimic the behavior of slicing lists. However, list slicing is lenient in the sense that any slice is considered valid. Even if the slice has out-of-bounds indices, Python will still return a result instead of raising an error. So there is no super easy way to check if the slice is requesting items that are not in the cache but would be if more results were downloaded. Instead I have to add logic to determine whether to fetch more results which gets complicated and error-prone when you consider all the possible cases of slices. It would require a lot of testing as well to ensure that the behavior matches list slicing.

We could also consider only implementing indexing for integer indices or a valid subset of slices, but I think this would lead to frustration/confusion when the behavior does not match that of list slicing.

I think the best option is to encourage users that want list-like behavior to create a list. I updated the docstring to demonstrate how to do that.

**Comment by mdehoon** [comment]
OK, then how about implementing `__getitem__` without using a cache?
I.e. `__getitem__` simply downloads and returns the requested records.
I am pretty much OK with this PR, but this:
```python
>>> results = list(islice(UniProt.search(query, batch_size=50), 50))
```
is a bit ugly.
It's pretty easy to normalize a slice. Something like this:
```python
        def __getitem__(self, key):
            if isinstance(key, slice):
                start, stop, step = slice.indices(key, length)   # where length is the number of records you got from the header
                if step != 1:
                    raise ValueError("only slices with step=1 are accepted")
                # and proceed to download and return records [start:stop]
            else:  # key is an integer
                # download and return the single record
```
If you prefer, we can also leave this for some other PR some other time.

**Comment by Will-Tyler** [comment]
The `indices` method contains most of the logic that I was trying to implement. With that method handy, I implemented indexing and slicing of the results, and now I am more confident that the code is easier to understand and maintain and is free of errors. The behavior should mimic that of indexing and slicing lists. Thanks for the suggestion!

**Comment by mdehoon** [comment]
@Will-Tyler  Thank you!

Did you see the "cursor" parameter on this page: https://www.uniprot.org/help/api_queries
I believe it lets you start downloading records from any start index. Would that simplify the code?

**Comment by Will-Tyler** [comment]
I am actually not sure how the cursor parameter works. Here is an example for the next page of results in the search for "Insulin AND (reviewed:true)": `https://rest.uniprot.org/uniprotkb/search?format=json&query=Insulin%20AND%20%28reviewed%3Atrue%29&cursor=1mkycb2xwxboua7hp3q76zzn28i0zp1bhe7n&size=25`. I am not sure how the value `1mkycb2xwxboua7hp3q76zzn28i0zp1bhe7n` was determined.

Also if I try to specify an integer position as the cursor, I get the following response: `{"url":"http://rest.uniprot.org/uniprotkb/search","messages":["Internal server error"]}`.

**Comment by mdehoon** [comment]
There is some explanation of the cursor parameter here:

https://www.uniprot.org/help/metalink
https://www.uniprot.org/help/pagination

but it looks like that this can only be used for iteration, not for random access. Then I think the current code is the perfectly fine. Thanks!

**Comment by Will-Tyler** [comment]
Thanks for reviewing and providing helpful comments!

**Pull request merged**